src/persist-txn/src/txns.rs

```diff
@@ -289,16 +284,104 @@ where
                 .insert(data_write.shard_id(), data_write);
         }
 
-        // Now that we've written the initialization into the txns shard, we
-        // know it's safe to make the data shard readable at `commit_ts`. We
-        // might crash between the txns write and this one, so `apply_le` also
-        // has to do the work just in case.
-        //
-        // TODO(txn): This will have to work differently once we plumb in real
-        // schemas.
-        self.apply_le(&register_ts).await;
+        Ok(())
+    }
+
+    /// Removes data shards from use with this txn set.
+    ///
+    /// The registration entry written to the txn shard is retracted. If it is
+    /// not possible to forget the data shard at the requested time, an Err will
+    /// be returned with the minimum time the data shards could be forgotten.
+    ///
+    /// This method is idempotent. Data shards currently forgotten at
+    /// `forget_ts` will not be forgotten a second time.
+    ///
+    /// **WARNING!** While a data shard is registered to the txn set, writing to
+    /// it directly (i.e. using a WriteHandle instead of the TxnHandle,
+    /// registering it with another txn shard) will lead to incorrectness,
+    /// undefined behavior, and (potentially sticky) panics.
+    #[instrument(level = "debug", skip_all, fields(ts = ?forget_ts, shard = %data_id))]
+    pub async fn forget(&mut self, forget_ts: T, data_id: ShardId) -> Result<(), T> {
+        let mut txns_upper = self
+            .txns_write
+            .shared_upper()
+            .into_option()
+            .expect("txns should not be closed");
+        loop {
+            self.txns_cache.update_ge(&txns_upper).await;
 
-        Ok(register_ts)
+            let (key, val) = C::encode(crate::TxnsEntry::Register(data_id));
+            let updates = if self.txns_cache.registered_at(&data_id, &txns_upper) {
+                vec![((&key, &val), &forget_ts, -1)]
+            } else {
+                // Never registered or already forgotten. This could change in
+                // `[txns_upper, forget_ts]` (due to races) so close off that
+                // interval before returning, just don't write any updates.
+                vec![]
+            };
+
+            // If the txns_upper has passed forget_ts, we can no longer write.
+            // Return success if we have nothing left to write or an Error if we
+            // do.
+            if forget_ts < txns_upper {
+                if updates.is_empty() {
+                    return Ok(());
```

Comment:
Like register, I'm slightly worried that this might break the guarantee of the function if the shard was forgotten at a timestamp `ts` s.t. `forget_st < ts < tx_upper`.

Reply:
Ditto